File: 10week/남규/1. Merge k Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<pair<int, ListNode*>> pq;
        
        for(auto list : lists){
            if(list == NULL) continue;
            pq.push({-list->val, list});
        }
        
        ListNode* prev = new ListNode(-1);
        ListNode* curr = NULL;
        
        while(!pq.empty()){
            int leastVal = -pq.top().first;
            ListNode* leastList  = pq.top().second;
            pq.pop();
            
            if(curr == NULL){
                curr = leastList;
                prev->next = curr;
            } else {
                curr->next = leastList;
                curr = curr->next;
            }
            
            if(leastList->next){
                pq.push({-leastList->next->val, leastList->next});
            }
        }
        return prev->next;
    }
};

File: 10week/홍준/1.merge-k-sorted-lists.py
## https://leetcode.com/problems/merge-k-sorted-lists/

from typing import List, Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        val_list = []
        for head in lists:
            while head:
                val_list.append(head.val)
                head = head.next
        if not val_list:
            return None
        val_list.sort()
        # print(val_list)
        head = tail = ListNode(val_list[0], None)
        for val in val_list[1:]:
            tail.next = ListNode(val, None)
            tail = tail.next
        return head
        